salesagility/SuiteCRM
# Hotfix 7.8.x tests

## Description




- Introduce State Saver ans State Checker
see more about how to use state saver ans state checker:


- Implementing state checker unit tests
see more about how to write phpunit tests: 
https://docs.suitecrm.com/developer/appendix-c---automated-testing/#_writing_and_running_tests

- Clean up tests folder

- Fixing test and related code base

- Setting up php unit code coverage

- Adding fancy output for unit test in terminal, per tests is visible now:
![image](https://user-images.githubusercontent.com/13831178/40056677-fc61fde6-5843-11e8-87c7-c00c366528af.png)

(TODO: this part needs to update documentation for SuiteCRM 7.8.x only)

## Motivation and Context


- CRM needs unit tests
- developers want to create and run tests and code coverage report

## How To Test This


- Install SuiteCRM

- run in terminal:
```
$ cd tests
$ ../vendor/bin/phpunit
```

- run tests separately:
`$ ../vendor/bin/phpunit [test sub-folder or test filename]`


Extra benefit for NetBeans users:

- Project setup:
![image](https://user-images.githubusercontent.com/13831178/40056323-f77a1648-5842-11e8-97de-1b5dbd480a6f.png)

- Run tests:
Right click on project name -> Test (Alt-F6)

- Collect Coverage:
Right click on project name -> Code Coverage -> Collect and Display Code Coverage
Run tests
See the covered and uncovered (green/red) code in your IDE:
![image](https://user-images.githubusercontent.com/13831178/40056877-90a8278c-5844-11e8-909e-429ff0cb0e0e.png)
See full Code Coverage Report for SuiteCRM in your IDE:
![image](https://user-images.githubusercontent.com/13831178/40056947-c1fc2e00-5844-11e8-9820-8b0f98932832.png)

## Types of changes


### Final checklist
